File: src/backend/GPTAgent.py
from openai import AzureOpenAI
from AgentSettings import AgentSettings
from Models import ChatMessage
import database as rep

class GPTAgent:
    def __init__(self, settings = None, client = None):

        if settings is None:
            settings = AgentSettings()

        if client is None:
            client = AzureOpenAI(azure_endpoint=settings.api_endpoint, api_key=settings.api_key, api_version=settings.api_version)

        self.settings : AgentSettings = settings
        self.client : AzureOpenAI = client
        self.get_context_delegate = None

    def process(self, user_name: str, user_id: str, prompt: str,max_tokens:int=500,temperature:float=0.3,context:str="") -> list:
        
        if self.get_context_delegate:
            context = self.get_context_delegate()

        completion = self.client.chat.completions.create(
                model=self.settings.model_deployment,
                messages=[
                    {
                        "role": "system",
                        "content": "You are an agent that can help answer questions about customers, products, and customer orders." ,
                    },
                    {
                        "role": "user",
                        "content": prompt +"Text: \"\"\"" + context + "\"\"\"",
                    }
                ],
                max_tokens=max_tokens,
                temperature=temperature
            )
        
        #return ('assistant',user_name,user_id,str(completion.choices[0].message.content))
        result = str(completion.choices[0].message.content)        
        output_messages = [
            ChatMessage(role='user',user_name=user_name,user_id=user_id,content=prompt,columns=[],rows=[]),
            ChatMessage(role='assistant',user_name=user_name,user_id=user_id,content=result,columns=[],rows=[])            
        ]
        return output_messages
    

File: src/csbackend/Agents/GPTAgent.cs
using Azure.AI.OpenAI;
using models;

namespace agents;

public class GPTAgent(AgentSettings? settings, OpenAIClient? client) : BaseAgent(settings, client)
{
    public override async Task<List<ChatMessage>> ProcessAsync(string user_name, string user_id, string prompt, int max_tokens = 500, float temperature = 0.3f, string context = "")
    {
        if (GetContent is not null)
        {
            context += GetContent();
        }

        var fullPrompt = $"{prompt}\nText: \"\"\"\n{context}\n\"\"\"";
        var completion = await AgentUtils.CallGPTAsync(Client, Settings.GPT_Model_Deployment_Name, fullPrompt, temperature, max_tokens);
        if (completion is not null)
        {
            var outputMessages = new List<ChatMessage>
                {
                    new("user", user_name, user_id, prompt, [], []),
                    new("assistant", user_name, user_id, completion, [],[])
                };
            return outputMessages;
        }
        else
        {
            return [];
        }
    }
}